### PR TITLE
implemented FetchPushDown and use it in QueryThenFetchConsumer

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    @Test
+    public void testWhereIsPushedDown() throws Exception {
+        QuerySpec qs = new QuerySpec();
+        qs.outputs(ImmutableList.<Symbol>of(REF_I, REF_A));
+        qs.where(WhereClause.NO_MATCH);
+
+        QuerySpec sub = FetchPushDown.pushDown(qs, TABLE_IDENT);
+
+        assertThat(sub.where(), is(WhereClause.NO_MATCH));
+        assertThat(qs.where(), is(WhereClause.MATCH_ALL));
```
Comment (bogensberger) [inline sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java]: why do you get a MATCH_ALL here, when you are setting the where clause to NULL in https://github.com/crate/crate/pull/2414/files#diff-a7666a18fefff52a10fced47a2cb4d41R107 ?

Comment (bogensberger) [inline sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java]: ah got it, you are returning in https://github.com/crate/crate/pull/2414/files#diff-a7666a18fefff52a10fced47a2cb4d41R81

File: sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+
```
Comment (bogensberger) [inline sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java]: please add a test with a more complex where clause, e.g.: something like:

where t.i < 8 and t.a > 3 and (t.i + t.a) < 10

Comment (mfussenegger) [inline sql/src/test/java/io/crate/planner/fetch/FetchPushDownTest.java]: why? There is no special push down logic for the where close. It is just set onto the sub query:

    sub.where(querySpec.where());
    querySpec.where(null);

**Comment by bogensberger** [comment]
ack

**Pull request merged**